internal/contentmanager/dynamicsource.go

```diff
@@ -0,0 +1,183 @@
+type DynamicInformerSource struct {
+	// DynamicInformerFactory is the dynamicinformer.DynamicSharedInformerFactory
+	// that is used to generate the informers configured on this sources startup
+	DynamicInformerFactory dynamicinformer.DynamicSharedInformerFactory
+	// GVR is the GroupVersionResource that this source is responsible
+	// for creating and configuring an informer for
+	GVR schema.GroupVersionResource
+	// Owner is the client.Object that owns the managed content that this
+	// source will be creating an informer to react to events for. This
+	// field is used to attempt to requeue the owning client.Object for
+	// reconciliation on a watch error after a previously successful sync
+	Owner client.Object
+	// Handler is the handler.EventHandler that is used to react to events
+	// received by the configured informer
+	Handler handler.EventHandler
+	// Predicates are the predicate.Predicate functions used to determine
+	// if a triggered event should be reacted to
+	Predicates []predicate.Predicate
+	// OnWatchError is the callback function that is used when the informer
+	// initially synced successfully and later encountered an error
+	OnWatchError func()
+	errOnce      sync.Once
+	errChan      chan error
+	syncedChan   chan struct{}
+	mu           sync.RWMutex
+	synced       bool
+	cancel       context.CancelFunc
+	ctx          context.Context
+	informer     informers.GenericInformer
+}
+
+var _ ReaderCloserSyncingSource = (*DynamicInformerSource)(nil)
+
+func (dis *DynamicInformerSource) String() string {
+	return "contentmanager.DynamicInformerSource"
+}
+
+func (dis *DynamicInformerSource) Start(ctx context.Context, q workqueue.RateLimitingInterface) error {
+	informerCtx, cancel := context.WithCancel(ctx)
+
+	dis.mu.RLock()
+	gInf := dis.DynamicInformerFactory.ForResource(dis.GVR)
+	dis.mu.RUnlock()
+
+	dis.mu.Lock()
+	dis.errChan = make(chan error, 1)
+	dis.syncedChan = make(chan struct{}, 1)
+	dis.informer = gInf
+	dis.cancel = cancel
+	dis.ctx = informerCtx
+	eventHandler := NewEventHandler[client.Object](ctx, q, dis.Handler, dis.Predicates)
+	dis.mu.Unlock()
+
+	// If we encounter an error during the watch we will:
+	// - Capture the error
+	// - cancel the informer
+	// requeuing of the ClusterExtension should happen by the
+	// WaitForSync function returning an error
+	// Only if we have successfully synced in the past should we
+	// requeue the ClusterExtension
+	sharedIndexInf := gInf.Informer()
+	err := sharedIndexInf.SetWatchErrorHandler(func(r *cgocache.Reflector, err error) {
+		dis.mu.Lock()
+		dis.errOnce.Do(func() {
+			dis.errChan <- err
+		})
+		dis.cancel()
+
+		if dis.synced {
+			// We won't be able to update the ClusterExtension status
+			// conditions so instead force a requeue if we
+			// have previously synced and then errored
+			q.Add(reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name: dis.Owner.GetName(),
+				},
+			})
+			dis.OnWatchError()
```

Comment:
We should pass a context here. Since the caller in this case can't handle errors they should be handled by the callback itself.